src/statefunctions/gamestates/chi.cc

```diff
@@ -2,53 +2,71 @@
 #include <cstdint>
 #include <iostream>
 #include <memory>
-#include <vector>
 
 #include "statefunctions/router.h"
 #include "statefunctions/stateutilities.h"
 #include "types/event.h"
 #include "types/gamestate.h"
+#include "types/hand.h"
+#include "types/meld.h"
 #include "types/pieces.h"
 #include "types/piecetype.h"
 #include "types/sets.h"
 #include "types/statefunction.h"
 
 namespace mahjong {
 namespace {
-Piece GetChiStart(const GameState& state, int player) {
-  if (CountPieces(state, player, state.pendingPiece - 2) > 0 &&
-      CountPieces(state, player, state.pendingPiece - 1) > 0) {
-    return state.pendingPiece - 2;
+Piece GetChiStart(const Hand& hand, Piece piece) {
+  if (CountPieces(hand, piece - 2) > 0 && CountPieces(hand, piece - 1) > 0) {
+    return piece - 2;
   }
-  if (CountPieces(state, player, state.pendingPiece - 1) > 0 &&
-      CountPieces(state, player, state.pendingPiece + 1) > 0) {
-    return state.pendingPiece - 1;
+  if (CountPieces(hand, piece - 1) > 0 && CountPieces(hand, piece + 1) > 0) {
+    return piece - 1;
   }
-  if (CountPieces(state, player, state.pendingPiece + 1) > 0 &&
-      CountPieces(state, player, state.pendingPiece + 2) > 0) {
-    return state.pendingPiece;
+  if (CountPieces(hand, piece + 1) > 0 && CountPieces(hand, piece + 2) > 0) {
+    return piece;
   }
   return kError;
 }
 
 std::unique_ptr<GameState> Chi(std::unique_ptr<GameState> state) {
+  if (Hand& hand = state->hands[state->currentPlayer];
+      hand.riichi && hand.discards_count == hand.riichiPieceDiscard) {
+    hand.riichiPieceDiscard++;
+  }
+
+  state->currentPlayer = state->lastCaller;
+  state->lastCall = state->turnNum;
+  state->concealedKan = false;
+  state->turnNum++;
+
+  Hand& hand = state->hands[state->currentPlayer];
```

Comment:
nit: is this technically a double array access with the if statement above?

Reply:
We update state->currentPlayer to state->lastCaller on line 38. So it's accessing a diff player now.

Reply:
ah! got it